tests/test_skvbc_fast_path.py

```diff
@@ -0,0 +1,158 @@
+
+        We send a new series of K/V writes and make sure they
+        have been processed using the slow commit path.
+
+        Finally we check if a known K/V has been executed.
+        """
+        trio.run(self._test_fast_to_slow_path_transition)
+
+    async def _test_fast_to_slow_path_transition(self):
+        for bft_config in bft.interesting_configs():
+            config = bft.TestConfig(n=bft_config['n'],
+                                    f=bft_config['f'],
+                                    c=bft_config['c'],
+                                    num_clients=bft_config['num_clients'],
+                                    key_file_prefix=KEY_FILE_PREFIX,
+                                    start_replica_cmd=start_replica_cmd)
+            with bft.BftTestNetwork(config) as bft_network:
+                await bft_network.init()
+                bft_network.start_all_replicas()
+
+                for _ in range(10):
+                    await bft_network.write_known_kv(self.protocol)
+
+                await bft_network.assert_fast_path_prevalent(as_of_seq_num=1)
+
+                unstable_replicas = list(set(range(0, config.n)) - {0})
+                bft_network.stop_replica(
+                    replica=random.choice(unstable_replicas))
+
+                for _ in range(10):
+                    key, val = await bft_network.write_known_kv(self.protocol)
+
+                await bft_network.assert_slow_path_prevalent(as_of_seq_num=10)
+
+                await bft_network.assert_kv_write_executed(self.protocol, key, val)
+
+    @unittest.skip("According to the SBFT spec, the fast path should tolerate up to c crashed nodes."
+                   "Enable the test once this resilience gets implemented.")
+    def test_fast_path_resilience_to_crashes(self):
+        """
+        In this test we check the fast path's resilience to the failure of up to "c" nodes.
+
+        As a first step, we bring down no more than c replicas.
+
+        Then we write a series of known K/V entries, making sure
+        we stay on the fast path.
+
+        Finally we check if a known K/V has been executed.
+        """
+        trio.run(self._test_fast_path_resilience_to_crashes)
+
+    async def _test_fast_path_resilience_to_crashes(self):
+        for bft_config in bft.interesting_configs(c_min=1):
+            config = bft.TestConfig(n=bft_config['n'],
+                                    f=bft_config['f'],
+                                    c=bft_config['c'],
+                                    num_clients=bft_config['num_clients'],
+                                    key_file_prefix=KEY_FILE_PREFIX,
+                                    start_replica_cmd=start_replica_cmd)
+            with bft.BftTestNetwork(config) as bft_network:
+                await bft_network.init()
+                bft_network.start_all_replicas()
+
+                unstable_replicas = list(set(range(0, config.n)) - {0})
```

Comment:
This type of list(set(range)) code is used in many places. We should abstract it to a shared function. Something like `unstable_replicas = set_subtract(config.n, {0})`. I don't necessarily like that name, but I've just been thinking about how we could reduce this duplication, which is somewhat hard to read. It can also be done in a separate PR. I'm just pointing it out here while thinking of it.

On the other hand, maybe it's better that it's explicitly done here this way. Any python programmer should grasp it.

Reply:
Indeed, it can be added to bft.py - will do it in a separate PR.